Python/flowgraph.c

```diff
@@ -1344,6 +1349,23 @@ apply_static_swaps(basicblock *block, int i)
                 return;
             }
         }
+        // The reordering is not safe if the two instructions to be swapped
+        // store to the same location, or if any intervening instruction stores
+        // to the same location as either of them.
+        int store_j = STORES_TO(block->b_instr[j]);
+        int store_k = STORES_TO(block->b_instr[k]);
+        if (store_j >= 0 || store_k >= 0) {
+            if (store_j == store_k) {
+                return;
+            }
+            for (int idx = j + 1; idx < k; idx++) {
+                int store_idx = STORES_TO(block->b_instr[idx]);
+                if (store_idx >= 0 && (store_idx == store_j || store_idx == store_k)) {
```

Comment:
How would it happen that there's a STORE_FAST* between j and k?

Reply:
This can happen with code like `a, a, b = x, y, z` as in some of the added tests. This compiles to `LOAD_FAST x; LOAD_FAST y; LOAD_FAST z; SWAP 3; STORE_FAST a; STORE_FAST a; STORE_FAST b`. Before this PR, `apply_static_swaps` would optimize that (ignoring the loads) to `STORE_FAST b; STORE_FAST a; STORE_FAST a` (swapping the first and third store), which is invalid because it reorders the two stores to `a`.

`k - j == n -1` here, where `n` is the oparg to the `SWAP`. So for a `SWAP 2`, `j` and `k` will be adjacent, but for `SWAP` with oparg > 2, there will be intervening instructions.